08_project/n0npax/pkg/puppy/store/type.go

```diff
@@ -0,0 +1,26 @@
+package store
+
+import (
+	"sync"
+
+	puppy "github.com/anz-bank/go-course/08_project/n0npax/pkg/puppy"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+//SyncStore sync.Map based type for storing puppies data
+type SyncStore struct {
+	sync.Map
+	sync.Mutex
+	total int
+}
+
+// MemStore map based type for storing puppies data
+type MemStore map[int]puppy.Puppy
+
+// LevelDBStore provides sync for leveldb
+type LevelDBStore struct {
+	ldb *leveldb.DB
+	sync.Mutex
+	total int
```

Comment:
This is a bit of a race condition. What if somebody else adds a puppy to your DB in another way?
I've had a bit of a look at the LevelDB API - not sure if there is a better solution - this might just be the best we can do....
Aside: total is a bit of a misleading name as it is not the total of all entries once you start deleting.
Maybe call it nextId or similar?

Reply:
`total` renamed to `nextID`.

I believe 1st part of the comment is overkill.
If someone is adding a puppy to any store in another way than API, there will be always race condition.
Let's say we're working with `mysql`. Someone can be connected directly to the InnoDB engine and skip most of the `mysql` synchronization. Besides this, most of the client libraries assume nobody modifies persistent layer in a hacky way in the midtime.